src/splitter.rs

```diff
@@ -15,8 +17,109 @@ impl Splitter {
             .collect()
     }
 
-    pub fn phrases(&self, _text: &str) -> Vec<Vec<u16>> {
-        todo!("Implement phrases extraction")
+    pub fn phrases(&self, text: &str) -> Vec<Vec<u16>> {
+        // Split into sentences using delimiters: . ? \n\n ;
+        let sentences = self.split_into_sentences(text);
+        
+        // Collect all phrases (word sequences) from all sentences
+        let mut all_phrases = BTreeSet::new();
+        
+        for sentence in &sentences {
+            let cleaned_words = self.clean_sentence(sentence);
+            if cleaned_words.len() >= 2 {
+                // Generate all substrings of length >= 2
+                for phrase in self.generate_substrings(&cleaned_words) {
+                    all_phrases.insert(phrase);
+                }
+            }
+        }
+        
+        // Build vocabulary from the same cleaned text that we use for phrases
+        let vocabulary = self.build_cleaned_vocabulary(text);
+        
+        // Convert phrases to indices
+        all_phrases.into_iter()
+            .map(|phrase| self.phrase_to_indices(&phrase, &vocabulary))
+            .collect()
+    }
+    
+    fn split_into_sentences(&self, text: &str) -> Vec<String> {
```

Comment:
Use chaining and pure transforms here. Use a functional style. Don't use mutation.